src/test/java/com/marklogic/kafka/connect/source/ConstraintStorageTest.java

```diff
@@ -0,0 +1,87 @@
+        );
+        List<SourceRecord> newRecords = task.poll();
+        Assertions.assertEquals(15, newRecords.size(), "All 15 initial records should be returned");
+        String previousMaxConstraintColumnValue = task.getPreviousMaxConstraintColumnValue();
+
+
+        ConstraintValueStore constraintValueStore = new MarkLogicConstraintValueStore(getDatabaseClient(), constraintStorageUri,
+            constraintColumnName, new HashMap<>());
+        String lastValue = constraintValueStore.retrievePreviousMaxConstraintColumnValue();
+        Assertions.assertEquals(previousMaxConstraintColumnValue, lastValue,
+            "The stored lastValue should equal the value from the Task");
+        PermissionsTester permTester = this.readDocumentPermissions(constraintStorageUri);
+        permTester.assertReadPermissionExists(roleName);
+        permTester.assertUpdatePermissionExists(roleName);
+        this.assertInCollections(constraintStorageUri, constraintStorageCollections.split(","));
+
+        String laterTime = "02:01:00";
+        loadSingleAuthorRowIntoMarkLogicWithCustomTime("firstHigherId", "6", laterTime, "firstHigherId");
+        loadSingleAuthorRowIntoMarkLogicWithCustomTime("secondHigherId", "7", laterTime, "secondHigherId");
+        loadSingleAuthorRowIntoMarkLogicWithCustomTime("lowestId", "0", laterTime, "lowestId");
+
+        newRecords = task.poll();
+        Assertions.assertEquals(2, newRecords.size(), "Only 2 of the 3 new records should be returned");
+        Assertions.assertTrue(newRecords.toString().contains("\"Medical.Authors.ID\":{\"type\":\"xs:integer\",\"value\":6}"));
+        Assertions.assertTrue(newRecords.toString().contains("\"Medical.Authors.ID\":{\"type\":\"xs:integer\",\"value\":7}"));
+
+
+        String newLastValue = constraintValueStore.retrievePreviousMaxConstraintColumnValue();
+        previousMaxConstraintColumnValue = task.getPreviousMaxConstraintColumnValue();
+        Assertions.assertEquals(previousMaxConstraintColumnValue, newLastValue,
+            "The stored lastValue should have been updated to the value from the second run of the task");
+
+
+        String evenLaterTime = "03:01:00";
+        loadSingleAuthorRowIntoMarkLogicWithCustomTime("thirdHigherId", "8", evenLaterTime, "thirdHigherId");
+        loadSingleAuthorRowIntoMarkLogicWithCustomTime("fourthHigherId", "9", evenLaterTime, "fourthHigherId");
+
+        storeBadConstraintState(constraintStorageUri);
+        String badLastValue = constraintValueStore.retrievePreviousMaxConstraintColumnValue();
+        Assertions.assertNull(badLastValue, "The retrieved last value should be null since the document is invalid");
+
+        newRecords = task.poll();
+        Assertions.assertEquals(20, newRecords.size(),
+            "Since the stored state document is invalid, all 21 records should have been returned");
+    }
+
+    public void storeBadConstraintState(String constraintStorageUri) {
```

Comment:
This should be private, just to make it clear it's not used anywhere else. Also, this is going to fail not because the data is invalid, but because the doc is lacking a read permission and the user won't see it. I think it'd be better, and less code, to just reuse the ConstraintValueStore but pass in a value that isn't valid. 

Actually - I just tried this, and it appears to be a different scenario. I tried this:

    constraintValueStore.storeConstraintState("invalid-value");

And the behavior I got instead was that no rows were returned. 

That seems unexpected to me - i.e. I would expect that if there's any problem with the constraint value document (it's missing, it's invalid, or the doc is valid JSON but the value is invalid), I would get the same behavior - which I think should be that all rows are returned. 

I think that's a good scenario to add, and I think both that and this scenario of "Doc is invalid" should be in separate test methods. It makes this method easier to understand (it's 60 LOC, so a little long already), and makes it easier to understand what happens for these two error scenarios. 

A 4th error scenario - try writing the doc with no permissions. That should fail, and we want to document what happens there.

Reply:
I'm not real sure I've covered what you're looking for. Maybe we should take a look at it together.